src/listwidgetbackend.cpp

```diff
@@ -42,10 +54,48 @@ std::uint32_t ListWidgetBackend::get_num_lines()
 	return num_lines;
 }
 
-void ListWidgetBackend::update_position(std::uint32_t pos, std::uint32_t scroll_offset)
+void ListWidgetBackend::invalidate_list_content(std::uint32_t line_count,
+	std::function<std::string(std::uint32_t, std::uint32_t)> get_line_method)
+{
+	line_cache.clear();
+	get_formatted_line = get_line_method;
+	num_lines = line_count;
+
+	on_list_changed();
+
+	render();
+}
+
+void ListWidgetBackend::update_position(std::uint32_t pos,
+	std::uint32_t new_scroll_offset)
+{
+	scroll_offset = new_scroll_offset;
+	form.set(list_name + "_pos", std::to_string(pos - scroll_offset));
+	form.set(list_name + "_offset", "0");
+
+	render();
+}
+
+void ListWidgetBackend::render()
 {
-	form.set(list_name + "_pos", std::to_string(pos));
-	form.set(list_name + "_offset", std::to_string(scroll_offset));
+	const auto viewport_width = get_width();
+	const auto viewport_height = get_height();
+	const auto visible_content_lines = std::min(viewport_height, num_lines - scroll_offset);
+
+	listfmt.clear();
+	for (std::uint32_t i = 0; i < visible_content_lines; ++i) {
+		const std::uint32_t line = scroll_offset + i;
+		std::string formatted_line = "NO FORMATTER DEFINED";
+		if (line_cache.count(line) >= 1) {
+			formatted_line = line_cache[line];
```

Comment:
Does this cache save us any work at all? As far as I understand, `invalidate_list_content` is called pretty much on every iteration of the `View::run()` (via `FormAction::prepare()`), so the cache ends up constantly invalidated. The cache probably helps in `ListWidgetBackend::update_position`, but positions are updated in response to keypresses, which means there will be another iteration of `View::run()` afterwards and the cache will get invalidated again.

I'm not asking for the cache to be removed; I'm just curious why it was introduced at all.

Reply:
The `invalidate_list_content()` calls are often guarded by an `if (do_redraw)`.
As far as I know, that flag is not set for regular movement commands.

> I'm not asking for the cache to be removed; I'm just curious why it was introduced at all.

If I remember correctly, it mainly had a big impact on update times when moving up or down by a single line (in that case, all lines except 1 are already in the cache).

Reply:
> As far as I know, that flag is not set for regular movement commands.

Seems to be so. Let's keep the cache then.